src/stdlib_math_diff.fypp

```diff
@@ -0,0 +1,102 @@
+#! Originally written in 2016 by Keurfon Luu (keurfonluu@outlook.com)
+#! https://github.com/keurfonluu/Forlab/blob/master/src/lib/forlab.f90#L2673
+
+#:include "common.fypp"
+#:set RI_KINDS_TYPES = REAL_KINDS_TYPES + INT_KINDS_TYPES
+submodule (stdlib_math) stdlib_math_diff
+
+    implicit none
+
+contains
+
+    #! `diff` computes differences of arrays of the ${t1}$ type.
+    
+    #:for k1, t1 in RI_KINDS_TYPES
+    pure module function diff_1_${k1}$(X, n, prepend, append) result(Y)
+        ${t1}$, intent(in) :: X(:)
+        integer, intent(in), optional :: n
+        ${t1}$, intent(in), optional :: prepend(:), append(:)
+        ${t1}$, allocatable :: Y(:)
+        integer :: size_prepend, size_append
+        integer :: n_, i
+
+        size_prepend = 0
+        size_append = 0
+        n_ = optval(n, 1)
+        if (present(prepend)) size_prepend = size(prepend) 
+        if (present(append)) size_append = size(append)
+        
+        if (size(X) + size_prepend + size_append <= n_) then
+            allocate(Y(0))
+            return
+        end if
+
+        if (size_prepend > 0) then
+            Y = [prepend, X]
+        else
+            Y = X
+        end if
+        if (size_append > 0) Y = [Y, append]
```

Comment:
Is it not possible to avoid this reallocation?
E.g.
1) allocate `Y` with the full size
2) copy all elements in their right place